x/mongo/driver/topology/server_test.go

```diff
@@ -309,6 +312,94 @@ func TestServer(t *testing.T) {
 			t.Fatal("client metadata not expected in heartbeat but found")
 		}
 	})
+	t.Run("heartbeat monitoring", func(t *testing.T) {
+		var publishedEvents []interface{}
+
+		serverHeartbeatStarted := func(e *event.ServerHeartbeatStartedEvent) {
+			publishedEvents = append(publishedEvents, *e)
+		}
+
+		serverHeartbeatSucceeded := func(e *event.ServerHeartbeatSucceededEvent) {
+			publishedEvents = append(publishedEvents, *e)
+		}
+
+		serverHeartbeatFailed := func(e *event.ServerHeartbeatFailedEvent) {
+			publishedEvents = append(publishedEvents, *e)
+		}
+
+		sdam := &event.ServerMonitor{
+			ServerHeartbeatStarted:   serverHeartbeatStarted,
+			ServerHeartbeatSucceeded: serverHeartbeatSucceeded,
+			ServerHeartbeatFailed:    serverHeartbeatFailed,
+		}
+
+		dialer := &channelNetConnDialer{}
+		dialerOpt := WithDialer(func(Dialer) Dialer {
+			return dialer
+		})
+		serverOpts := []ServerOption{
+			WithConnectionOptions(func(connOpts ...ConnectionOption) []ConnectionOption {
+				return append(connOpts, dialerOpt)
+			}),
+			withMonitoringDisabled(func(bool) bool { return true }),
+			WithServerMonitor(func(*event.ServerMonitor) *event.ServerMonitor { return sdam }),
+		}
+
+		s, err := NewServer(address.Address("localhost:27017"), primitive.NewObjectID(), serverOpts...)
+		if err != nil {
+			t.Fatalf("error from NewServer: %v", err)
+		}
+
+		// set up heartbeat connection, which doesn't send events
+		_, err = s.check()
+		assert.Nil(t, err, "check error: %v", err)
+
+		channelConn := s.conn.nc.(*drivertest.ChannelNetConn)
+		_ = channelConn.GetWrittenMessage()
+
+		t.Run("success", func(t *testing.T) {
+			publishedEvents = nil
+			// do a heartbeat with a non-nil connection
+			if err = channelConn.AddResponse(makeIsMasterReply()); err != nil {
+				t.Fatalf("error adding response: %v", err)
+			}
+			_, err = s.check()
```

Comment:
I described this issue in Slack a bit, but I'll summarize here so it's recorded in the PR. All calls to `check` have to then call `GetWrittenMessage` to clear the connection's write buffer. If this isn't done, the next `failure` test will fail to write anything to the buffer because it's full.

Reply:
done